[SYCL] Host pipe runtime implementation

zibaiwan opened the pull request:
Disclaimer:

This work is a continuation of a previous approved Sherry's PR: https://github.com/intel/llvm/pull/5766 and her draft work https://github.com/intel/llvm/pull/5894.

We are implementing Hostpipes based on the Spec [here](https://github.com/rho180/llvm/blob/rho180-pipe-design/sycl/doc/design/HostPipes.md). OpenCL spec is [here](https://github.com/intel-sandbox/ip-authoring-specs/blob/MJ_ChangeDocs4/Pipe/Spec/cl_intel_host_pipe_symbol.asciidoc).

The following is the outline of the design.

1.  The host pipe properties need to be added to the device image (probably similar to the previous change here https://github.com/intel/llvm/commit/a9ad3af71cbdd5ebc7b7efeb35761aa8ea1a49d7)

2.The frontend calls the [registration of device image, maybe similar to this code](https://github.com/intel/llvm/blob/af858c74e4be95b2163ce6ba545ce588ff0aca4c/sycl/source/detail/program_manager/program_manager.cpp#L2014): which is where the host pipe information is available. This is where the mapping from host pipe name to host pipe pointer is extracted.

3. The frontend also calls `host_pipe_map::add` to register/initialize the host pipes. This is the new function. We are not sure about the ordering of registration of device image / registration of host pipe. for which ever one that comes later, it need to initialize the remaining attribute of host pipe (such as its properties).

4. The opencl runtime will need to get a cl_program object, which is typically not available until the first kernel launch. To get a program object early on, the host pipe name/pointer to device image mapping is cached during registration. And when the specific host pipe is needed, build the program and get its ocl runtime representation. This is done in the first couple commits.

5. Since a host pipe read/write need to depend on other write operation finishing before it (including the inter kernel write). This means the pipe needs to know the dependency of kernel execution. For this reason, the host pipe read and write ocl function cannot be called with no dep event. therefore, it is implemented with handler , which is aware of the event that it is supposed to wait upon. This is done in the "Register new command group .." commit.

6. Unit test
    - mock a fake device image.
    - register the fake device image
    - register fake pipe with some name you specified
    - fake the opencl functionality, this can be done with unittest::PiMock::redefine

asudarsa reviewed (commented):
I do not see any changes relevant to dpcpp-tools. 

Thanks

steffenlarsen commented:
Converted to draft to avoid accidental merge.

steffenlarsen commented on sycl/include/sycl/ext/intel/experimental/host_pipes.hpp:
```diff
@@ -0,0 +1,93 @@
+          "sycl-host-pipe",
+          nullptr)]] [[__sycl_detail__::
+                           host_pipe]] [[__sycl_detail__::
+                                             global_variable_allowed]] // may
```
To my knowledge it should be possible to do multiple attributes in the same `[[...]]`, which may make the formatting here a little more readable.

steffenlarsen commented on sycl/include/sycl/ext/intel/experimental/host_pipes.hpp:
```diff
@@ -0,0 +1,93 @@
+  static constexpr int32_t min_cap =
+      _propertiesT::template has_property<min_capacity_key>()
+          ? _propertiesT::template get_property<min_capacity_key>().value
+          : 0;
```
Does this work correctly when there is no `min_capacity` property in the list? To my knowledge the ternary operator isn't `constexpr` so it will try to evaluate both sides. Alternatively you can make a trait in the `detail` namespace and specialize the value of that to whether or not the property is there.

rho180 replied:
@steffenlarsen I'm working on combining the class in this header into ext::intel::experimental::pipes for Zibai. I'm not sure I understand your trait suggestion, could you elaborate? I guess another option is to use a lambda to evaluate this (as well as the other properties which all need to be passed) but that could be a little unwieldy.

steffenlarsen replied:
You could add something like
```c++
namespace detail {
template <typename Properties, typename PropertyKey, typename Cond = void>
struct ValueOrDefault {
  template <typename ValT> static constexpr ValT get(ValT Default) {
    return Default;
  }
};

template <typename Properties, typename PropertyKey>
struct ValueOrDefault<
    Properties, PropertyKey,
    std::enable_if_t<
        ext::oneapi::experimental::is_property_list_v<Properties> &&
        Properties::template has_property<PropertyKey>()>> {
  template <typename ValT> static constexpr ValT get(ValT) {
    return Properties::template get_property<PropertyKey>().value;
  }
};
} // namespace detail
```
then you can use it here as
```c++
 static constexpr int32_t min_cap = detail::ValueOrDefault<_propertiesT, min_capacity_key>::get<int32_t>(0);
```

steffenlarsen commented on sycl/include/sycl/ext/intel/experimental/host_pipes.hpp:
```diff
@@ -0,0 +1,93 @@
+} // namespace intel
+} // namespace ext
+} // __SYCL_INLINE_VER_NAMESPACE(_V1)
+} // namespace sycl
```
```suggestion
} // namespace sycl

```

steffenlarsen commented on sycl/include/sycl/handler.hpp:
```diff
@@ -2483,6 +2492,16 @@ class __SYCL_EXPORT handler {
+  /// The size of returned type for each read.
+  size_t HostPipeTypeSize = 0;
+  /// If the pipe operation is read or write, 1 for read 0 for write.
+  bool HostPipeRead = true;
```
To avoid ABI break in handler (sycl/test/abi/layout_handler.cpp) these should be moved to `handler_impl` instead.

zibaiwan replied:
Thanks @steffenlarsen. Since this is an experimental API (in sycl\include\sycl\ext\intel\experimental\pipes.hpp), can we get an exception?

Referring to ABI guideline: Features clearly marked as experimental are considered as an exception to this guideline.

steffenlarsen replied:
Sadly we can't make an exception because handler isn't experimental and this affects the ABI even if the user isn't using the extension.

zibaiwan replied:
Thanks @steffenlarsen for the explanation. Totally understand! I will look into the `handler_impl` to do the move, hopefully that is not complicated.

zibaiwan replied:
Thanks Steffen. I have moved the code into `handler_impl`.

steffenlarsen commented on sycl/source/detail/host_pipe.cpp:
```diff
@@ -0,0 +1,70 @@
+
+template <class _name, class _dataT, typename _propertiesT>
+_dataT host_pipe<_name, _dataT, _propertiesT>::read(queue &q,
+                                                    memory_order order) {
```
This won't work. The templates need to be resolvable at compile-time, which they won't be if they are in the runtime library source files. Instead you can have a templateless base class for `host_pipe` and define the parts that need library-source-only features here, e.g. the `ProgramManager` instance.

zibaiwan replied:
Thanks Steffen. I have fixed the code accordingly.

steffenlarsen commented on sycl/source/detail/host_pipe.cpp:
```diff
@@ -0,0 +1,70 @@
+} // namespace intel
+} // namespace ext
+} // __SYCL_INLINE_VER_NAMESPACE(_V1)
+} // namespace sycl
```
```suggestion
} // namespace sycl

```

steffenlarsen commented on sycl/unittests/pipes/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+)
+
+add_dependencies(PipeTests sycl)
+target_include_directories(PipeTests PRIVATE SYSTEM ${sycl_inc_dir})
```
```suggestion
target_include_directories(PipeTests PRIVATE SYSTEM ${sycl_inc_dir})

```

zibaiwan commented:
Rob and I are still doing some code/comments cleanup, but the functionality part is ready for review. Thanks!

steffenlarsen reviewed (changes requested):
Creating a blocking review to avoid this being merged unintentionally with the current changes to buildbot/dependency.py

zibaiwan commented:
/testwin

zibaiwan commented:
/testwin

zibaiwan commented:
> Creating a blocking review to avoid this being merged unintentionally with the current changes to buildbot/dependency.py

Thanks Steffen. The change has been merged into the official header repo, so I have updated the commit to the latest one in the config file. Please take a look!

zibaiwan commented on sycl/include/sycl/ext/intel/pipes.hpp:
```diff
@@ -52,7 +52,7 @@ template <class _name, class _dataT, int32_t _min_capacity = 0> class pipe {
     throw sycl::exception(
         sycl::make_error_code(sycl::errc::feature_not_supported),
-        "Pipes are not supported on a host device.");
+        "Device-side API are not supported on a host device. Please use host-side API instead..");
```
I have a question, @steffenlarsen, do you know if it's possible to make this a compile time error instead of a runtime error when the user is trying to call a device-side API on the host?

steffenlarsen replied:
It is not possible as both paths are compiled inside a kernel. A `sycl::exception` is the best option here.

sarnex reviewed (approved):
esimd emulator rb

zibaiwan commented:
Hi, @intel/dpcpp-l0-pi-reviewers, @intel/dpcpp-devops-reviewers, when you get a chance, could you please take a look at this PR? If possible, I am hoping to get this merged before the auto cutoff date for this release. Thank you very much!

zibaiwan commented:
Noticed this CI failure in L0 test.

```event_profiling_info.cpp.tmp.out: /__w/llvm/llvm/llvm_test_suite/SYCL/Basic/event_profiling_info.cpp:31: bool verifyProfiling(event): Assertion `Submit <= Start' failed.
```

This test was passing earlier and the new change should have nothing to do with this test. The test hasn't been updated for a whil. Document it here, but likely not related to my change.

steffenlarsen commented on sycl/include/sycl/handler.hpp:
```diff
@@ -2646,6 +2646,14 @@ class __SYCL_EXPORT handler {
+  /// /// \param Size the size of data getting read back / to. \param Blocking
+  /// if read/write opeartion is blocking \param Read 1 for read, 0 for write
+  void read_write_host_pipe(const std::string &Name, void *Ptr, size_t Size,
+                            bool Block, bool Read);
```
I don't see this function described in the [sycl_ext_intel_dataflow_pipes extension](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_intel_dataflow_pipes.asciidoc). Is there a separate PR adding it?

Also, I believe adding new members to existing SYCL interfaces mandates that you prefix them with `ext_intel_` (for `ext::intel` extensions). `device_global` was an exception as it added overloads to `memcpy`/`copy`.

zibaiwan replied:
Thanks Steffen. I am changing the name accordingly. Yes, there will be a PR making the change to the Spec. There was one here: https://github.com/intel/llvm/pull/5838/files, but it's outdated. I need to update it and create a new PR for it.

zibaiwan replied:
I am keeping the name as ext_intel_read_write_host_pipe instead of ext_intel_readWriteHostPipe for consistency.

steffenlarsen replied:
It would be best to have that up ASAP. Normally we would need that reviewed first before adding new interfaces.

> I am keeping the name as ext_intel_read_write_host_pipe instead of ext_intel_ReadWriteHostPipe for consistency.

Yes, `ext_intel_read_write_host_pipe` would be more in line with other SYCL interfaces.

zibaiwan replied:
Thanks @steffenlarsen. Yes, I am preparing the Spec review PR ASAP.

zibaiwan replied:
@steffenlarsen , I have just created a draft [PR ](https://github.com/intel/llvm/pull/8789)for the Spec update which includes the `ext_intel_read_write_host_pipe` function. I am still working on some wording on the Spec change.

steffenlarsen commented on sycl/source/detail/pipes.cpp:
```diff
@@ -0,0 +1,29 @@
+} // namespace intel
+} // namespace ext
+} // __SYCL_INLINE_VER_NAMESPACE(_V1)
+} // namespace sycl
```
```suggestion
} // namespace sycl

```

steffenlarsen commented on sycl/source/detail/program_manager/program_manager.cpp:
```diff
@@ -1312,6 +1311,40 @@ void ProgramManager::addImages(pi_device_binaries DeviceBinary) {
+          assert(HostPipeInfo.size() >= NumPropertySizeBytes + NumTypeBytes &&
+                 "Unexpected property size");
+          auto TypeSize = *reinterpret_cast<const std::uint32_t *>(
+              &HostPipeInfo[NumPropertySizeBytes]);
```
I would be better to use the `dropBytes` and `consume` member functions on `ByteArray` here to avoid strict aliasing violations. You can see these being used a couple lines further up for `DeviceGlobalInfo`.

steffenlarsen reviewed (changes requested):
Could you please go through the changes and ensure that variable naming is consistent throughout? Preferably it would adhere to https://llvm.org/docs/CodingStandards.html, but as a rule of thumb you should follow the style in the surrounding code. Public SYCL interfaces are of course excluded.

zibaiwan commented on llvm/lib/SYCLLowerIR/HostPipes.cpp:
```diff
@@ -0,0 +1,97 @@
+/// The function gets this value from the LLVM IR attribute \c
+/// sycl-unique-id.
+///
+/// @param GV [in] Device Global variable.
```
Thanks @steffenlarsen . @rho180 , can you please take a look at this specific comment?

zibaiwan replied:
Thanks @rho180 for working on this! @steffenlarsen , please take a look when you have a chance. Thank you very much!

zibaiwan commented on sycl/include/sycl/ext/intel/experimental/pipes.hpp:
```diff
@@ -11,29 +11,135 @@
+    return Properties::template get_property<PropertyKey>().value;
+  }
+};
+} // namespace detail
```
Thanks @steffenlarsen. I moved `ValueOrDefault` to sycl/include/sycl/ext/oneapi/properties/properties.hpp instead of property_utils.hpp. Because the `is_property_list_v` used in the `ValueOrDefault` is defined in the properties.hpp, if I move `ValueOrDefault` to `property_utils.hpp`, I would have to include the `properties.hpp` in the `property_utils.hpp`, however, because of `properties.hpp` already includes `property_utils.hpp`, it creates circular dependency between  `properties.hpp` and `properties_utils.hpp`.

Please let me know if this make senses and I missed anything.

steffenlarsen replied:
I see. Another option is to move into another level of utility headers, but since this is the only implementation using it right now I am okay with leaving it as-is for now.

zibaiwan commented on sycl/include/sycl/ext/intel/experimental/pipes.hpp:
```diff
@@ -11,29 +11,135 @@
+                     memory_order order = memory_order::seq_cst) {
+    const device Dev = q.get_device();
+    bool IsPipeSupported =
+        Dev.has_extension("cl_intel_program_scope_host_pipe");
```
Creating an issue placeholder for this: https://github.com/intel/llvm/issues/8771, we can do this as a follow-up.

zibaiwan commented on sycl/include/sycl/ext/intel/experimental/pipe_properties.hpp:
```diff
@@ -0,0 +1,190 @@
+      ready_latency_key, std::integral_constant<int, Latency>>;
+};
+
+struct bits_per_symbol_key {
```
Those struct names are matching public SYCL interfaces: https://github.com/intel/llvm/pull/5839/files

zibaiwan commented:
> Could you please go through the changes and ensure that variable naming is consistent throughout? Preferably it would adhere to https://llvm.org/docs/CodingStandards.html, but as a rule of thumb you should follow the style in the surrounding code. Public SYCL interfaces are of course excluded.

@steffenlarsen , thank you very much for your review. I have reviewed the code and fixed the variable naming accordingly, except the excluded public interfaces. I have addressed most of the comments except one Rob will be looking at.

aelovikov-intel reviewed (approved):
buildbot/dependency.py LGTM for devops.

I'm pretty sure having it in a separate PR would have been approved much faster.

zibaiwan commented:
> buildbot/dependency.py LGTM for devops.
> 
> I'm pretty sure having it in a separate PR would have been approved much faster.

Thank you very much. I will definitely try to do it in a separate PR next time. The reason I combined them together is because the runtime code needs the definition of the new header, and the header wasn't merged until 3 days ago, without that change, this PR CI compile will always fail. I could separate into a second PR once that header got merged into the official repo though. I will do that next time.

smaslov-intel commented on sycl/include/sycl/detail/pi.h:
```diff
@@ -1925,6 +1927,55 @@ pi_result piextEnqueueDeviceGlobalVariableRead(
+/// @param event returns an event object that identifies this read / write
+/// command and can be used to query or queue a wait for this command to
+/// complete.
+__SYCL_EXPORT pi_result piextEnqueueReadHostPipe(
```
@kbenzie: please take this change to Unified Runtime

smaslov-intel replied:
@zibaiwan : please update PI API version here https://github.com/intel/llvm/pull/7468/files#diff-952d2515bc7a8c91d37420f519ff726c0e6ce71f5a81bfd98e5c6fcfcaa3ada0R78-R82

zibaiwan replied:
@smaslov-intel , thank you very much for your review. I have update the PI version in the pi.h. Sorry I missed this earlier. Please take a look!

kbenzie replied:
@smaslov-intel tracking this in https://github.com/oneapi-src/unified-runtime/issues/400

zibaiwan commented:
Windows run failed due to missing symbol in the windows dump after merged with sycl branch.

```
There are new symbols in the new library. It is a non-breaking change. Refer to sycl/doc/developer/ABIPolicyGuide.md for further instructions.
The following symbols are new to the object file:

?setKernelCacheConfig@handler@_V1@sycl@@AEAAXW4_pi_kernel_cache_config@@@Z
```

@againull , it seems to be introduced by your #8597. I see you only updated the linux symbol dump. I am new contributing to this repo, doesn't the windows dump need to be updated as well?

zibaiwan commented:
@steffenlarsen , thanks for review this PR. All comments have been addressed or left as follow-ups in issue #8771, please take a look when you have a chance. Your help is really appreciated!

zibaiwan commented:
/testwin

zibaiwan commented:
> /testwin

Windows build check-sycl test still failed with same error. It seems like the windows test doesn't merge my PR with the latest sycl branch, but just use my branch directly so it's still missing the symbol in the dump file. Given it's the only test failing, so I would consider everything is passing.

zibaiwan commented:
/testwin

steffenlarsen commented on sycl/include/sycl/ext/intel/experimental/pipes.hpp:
```diff
@@ -11,29 +11,110 @@
+    event E = Q.submit([=](handler &CGH) {
+      CGH.ext_intel_read_host_pipe(PipeName, DataPtr, sizeof(_dataT) /* non-blocking */);
+    });
+    E.wait();
```
What is the point of it being non-blocking if we immediately wait for it here?

steffenlarsen replied:
Nevermind. I misunderstood the point of the blocking behavior.

zibaiwan replied:
@steffenlarsen 
This blocking parameter is used by the lower OpenCL function. If this is set to be non-blocking, the OpenCL function will not wait forever for the incoming data, and send event complete message right away regardless whether there is any data to read. If it's blocking, the opencl function will wait forever until there is data to read.

zibaiwan commented:
@steffenlarsen, thanks for the suggestion. I have made the new handler function private and the tests passed locally.

zibaiwan commented:
/testwin

zibaiwan commented:
Running the test on windows to make sure everything looks good on windows too.

steffenlarsen reviewed (approved):
Thank you, @zibaiwan. I think the final result it looking great!

zibaiwan commented:
/testwin

zibaiwan commented:
/testwin

zibaiwan commented:
/testwin

zibaiwan commented:
/testwin

zibaiwan commented:
/testwin

zibaiwan commented:
/testwin

zibaiwan commented:
/testwin

zibaiwan commented:
CI Status from the earlier commit: [839d7ee] (https://github.com/intel/llvm/pull/7468/commits/839d7eee193d306f35875d3c0306a297dda3a6d9) is valid for this PR as the following two are just two empty commits.

zibaiwan commented:
The CI for the latest code [commit ](https://github.com/intel/llvm/commit/839d7eee193d306f35875d3c0306a297dda3a6d9)passed: https://github.com/intel/llvm/actions/runs/4556840235/jobs/8043028160, except for the known failures. @steffenlarsen , this PR is ready to be merged now. Thank you very much for your help.

steffenlarsen commented:
Failed Tests (2):
  SYCL :: AtomicRef/atomic_memory_order_acq_rel.cpp - Fixed in https://github.com/intel/llvm/pull/8860
  SYCL :: GroupAlgorithm/reduce_sycl2020.cpp - Reported in https://github.com/intel/llvm/issues/8847
